firedrake/assemble.py

```diff
@@ -17,49 +17,61 @@
-         assembled -- either as a monolithic matrix ('aij' or 'baij'), a block matrix
-         ('nest'), or left as a :class:`.ImplicitMatrix` giving matrix-free
-         actions ('matfree').  If not supplied, the default value in
-         ``parameters["default_matrix_type"]`` is used.  BAIJ differs
-         from AIJ in that only the block sparsity rather than the dof
-         sparsity is constructed.  This can result in some memory
-         savings, but does not work with all PETSc preconditioners.
-         BAIJ matrices only make sense for non-mixed matrices.
-    :arg sub_mat_type: (optional) string indicating the matrix type to
-         use *inside* a nested block matrix.  Only makes sense if
-         ``mat_type`` is ``nest``.  May be one of 'aij' or 'baij'.  If
-         not supplied, defaults to ``parameters["default_sub_matrix_type"]``.
-    :arg appctx: Additional information to hang on the assembled
-         matrix if an implicit matrix is requested (mat_type "matfree").
-    :arg options_prefix: PETSc options prefix to apply to matrices.
+        a :class:`~slate.TensorBase` expression.
+    :arg tensor: (optional) an existing tensor object to place the result in.
+    :arg bcs: (optional) an iterable of boundary conditions to apply.
+    :kwarg form_compiler_parameters: (optional) a dict of parameters to pass to
+        the form compiler. Ignored if not assembling a
+        :class:`~ufl.classes.Form`. Any parameters provided here will be
+        overridden by parameters set on the :class:`~ufl.classes.Measure` in the
+        form.  For example, if a ``quadrature_degree`` of 4 is
+        specified in this argument, but a degree of 3 is requested in
+        the measure, the latter will be used.
+    :kwarg mat_type: (optional) string indicating how a 2-form (matrix) should be
+        assembled -- either as a monolithic matrix (``"aij"`` or ``"baij"``),
+        a block matrix (``"nest"``), or left as a :class:`.ImplicitMatrix` giving
+        matrix-free actions (``'matfree'``).  If not supplied, the default value in
+        ``parameters["default_matrix_type"]`` is used.  BAIJ differs
+        from AIJ in that only the block sparsity rather than the dof
+        sparsity is constructed.  This can result in some memory
+        savings, but does not work with all PETSc preconditioners.
+        BAIJ matrices only make sense for non-mixed matrices.
+    :kwarg sub_mat_type: (optional) string indicating the matrix type to
+        use *inside* a nested block matrix.  Only makes sense if
+        ``mat_type`` is ``nest``.  May be one of ``"aij"`` or ``"baij"``.  If
+        not supplied, defaults to ``parameters["default_sub_matrix_type"]``.
+    :kwarg appctx: (optional) additional information to hang on the assembled
+        matrix if an implicit matrix is requested (mat_type ``"matfree"``).
+    :kwarg options_prefix: (optional) PETSc options prefix to apply to matrices.
```

Comment:
this docstring is missing a `:returns:`

Reply:
It's particularly important too since what you get from `assemble` is not exactly intuitive. Obviously will improve when we have UFL cofunctions and matrices.

Reply:
There is quite a lot of detail in the bottom part of the docstring about the various things that get returned from this function so I've lazily just put `:returns: See below.`.